### PR TITLE
improve string.splitLines()

### PR DESCRIPTION
**Pull request by WalterBright**
Was using `splitLines()` and ran into some problems. Fixes:

1. improved documentation
2. eliminated excessive and quite unnecessary decoding - only LINESEP and PARASEP ever need decoding, and those only for UTF-8 (never ever need to decode surrogate pairs)
3. no longer throws on invalid UTF - current practice is to ignore invalid UTF if possible, and substitute in 'Replacement Character' 0xFFFD if not. This change simply ignores invalid UTF. UTF validation should be done separately by the user if desired, not on every string operation. Throwing wasn't documented for `splitLines()` anyway, so this should be ok.
4. eliminates memory allocation that `decode` may do (for its exceptions)
5. reduced number of variables in critical loop, should result in better register allocation
6. not benchmarked, but this should be significantly faster, important because this is a heavily used function

There are more issues with splitLines(), such as it does not accept input ranges, and its memory allocation could be eliminated if it exposed a range as output, but these should be addressed later.

### CODE REVIEW

**Comment by yebblies** [comment]
> 3. no longer throws on invalid UTF - current practice is to ignore invalid UTF if possible, and substitute in 'Replacement Character' 0xFFFD if not. This change simply ignores invalid UTF. UTF validation should be done separately by the user if desired, not on every string operation. Throwing wasn't documented for splitLines() anyway, so this should be ok.

Could you document and add a test for the new behavior?

**Comment by bearophile** [comment]
@WalterBright:
> this is a heavily used function

splitLines is used, but the most commonly used function of this kind is probably File.byLine, that currently is slow and in need of performance improvements (See also: https://issues.dlang.org/show_bug.cgi?id=11810 ).

File: std/string.d [hunk 1]
```diff
@@ -1897,46 +1897,78 @@ S capitalize(S)(S s) @trusted pure
     $(D "\r\n"), $(XREF uni, lineSep), and $(XREF uni, paraSep) as delimiters.
     If $(D keepTerm) is set to $(D KeepTerminator.yes), then the delimiter
     is included in the strings returned.
-  +/
+
+  Params:
+    s a string of $(D chars), $(D wchars), or $(D dchars)
```
Comment (andralex) [inline std/string.d]: Aren't you missing the `=` signs here?

File: std/string.d [hunk 2]
```diff
@@ -1897,46 +1897,78 @@ S capitalize(S)(S s) @trusted pure
...
     {
-        immutable c = decode(s, nextI);
-
-        if (c == '\r' || c == '\n' || c == lineSep || c == paraSep)
+        switch (s[i])
         {
-            immutable isWinEOL = c == '\r' && i + 1 < s.length && s[i + 1] == '\n';
-            auto iEnd = i;
+            case '\n':
+                retval.put(s[iStart .. i + (keepTerm == KeepTerminator.yes)]);
```
Comment (andralex) [inline std/string.d]: love this

File: std/string.d [hunk 3]
```diff
@@ -1897,46 +1897,78 @@ S capitalize(S)(S s) @trusted pure
...
+                iStart = i + 1;
+                break;
+
+            case '\r':
+                if (i + 1 < s.length && s[i + 1] == '\n')
```
Comment (andralex) [inline std/string.d]: This is a good candidate for bounds check elimination. Do we do it?

File: std/string.d [hunk 4]
```diff
@@ -1897,46 +1897,78 @@ S capitalize(S)(S s) @trusted pure
...
+                    ++i;
+                }
+                else
+                {
+                    goto case '\n';
+                }
+                break;
 
-            if (keepTerm == KeepTerminator.yes)
+            static if (s[i].sizeof == 1)
```
Comment (andralex) [inline std/string.d]: Don't we also need to do something for UTF-16?

File: std/string.d [hunk 5]
```diff
@@ -1897,46 +1897,78 @@ S capitalize(S)(S s) @trusted pure
...
             {
-                iEnd = isWinEOL? nextI + 1 : nextI;
+                /* Manually decode:
+                 *  lineSep is E2 80 A8
+                 *  paraSep is E2 80 A9
+                 */
+                case 0xE2:
+                    if (i + 2 < s.length &&
+                        s[i + 1] == 0x80 &&
+                        (s[i + 2] == 0xA8 || s[i + 2] == 0xA9)
```
Comment (andralex) [inline std/string.d]: `s[i + 2].among(0xA8, 0xA9)`

**Comment by andralex** [comment]
you have a fallthrough error in the autotester

File: std/string.d (hunk 3 above)
Comment (WalterBright) [inline std/string.d]: We should, but I haven't checked that case.

File: std/string.d (hunk 4 above)
Comment (WalterBright) [inline std/string.d]: No. LINESEP and PARASEP are encoded as single code units in UTF-16.

File: std/string.d (hunk 5 above)
Comment (WalterBright) [inline std/string.d]: I'm afraid that the way among() is implemented it won't be inlined or optimized well.

**Comment by WalterBright** [comment]
all comments are addressed.

File: std/string.d (hunk 5 above)
Comment (andralex) [inline std/string.d]: urgh

**Comment by andralex** [comment]
Auto-merge toggled on

**Pull request merged**